working/macros/feature-specification.md

```diff
@@ -304,20 +304,109 @@ class, allowing you to build up your code fragments however you like.
+macro is running (lets say its exported by the macro library), that identifier
+could be shadowed by another identifier in the library.
 
-We do have the start of something like this already available in the `builder`
-api - these have APIs to get a reference to a `Type` object. We will want to add
-the ability to do the same for any arbitrary identifier, and then the ability
-to emit references to these inside of [Code][] objects.
+To enable a macro to safely emit a reference to a known identifier, there is
+a `ResolvedIdentifier` subtype of `Code`. This class takes both a simple name
+for the identifier (no prefix allowed), as well as a library uri, where that
+identifier should be looked up.
+
+The generated code should be equivalent to adding a new import to the library,
+with the specified uri and a unique prefix. In the code the identifier will be
+emitted with that unique prefix followed by its simple name.
+
+Note that technically this allows macros to add references to libraries that
+the macro itself does not depend on, and the users application also may not
+depend on. This is discouraged, but not prevented, and should result in an error
+if it happens.
+
+### Generated declarations
+
+A key use of macros is to generate new declarations, and hand-authored code may
+refer to them — it may call macro-generated functions, read macro-generated
+fields, construct macro-generated classes, etc. This means that before macros
+are applied, code may contain identifiers that cannot be resolved. This is not
+an error. Any identifier that can't be resolved before the macro is applied is
+allowed to be resolved to a macro-produced identifier after macros are applied.
+
+Macros are not permitted to introduce declarations that directly conflict with
+existing declarations in the same library. These rules are the same as if the
+code were hand written.
+
+Macros may also end up adding new declarations which shadow existing symbols in
+the library, but don't directly conflict. In this case we want to ensure that
+the intent of any user written code is always clear. Consider the following
+example:
+
+```dart
+int get x => 1;
+
+@generateX 
+class Bar {
+  // Generated: int get x => 2;
+
+  // Should this return the top level `x`, or the generated instance getter?
+  int get y => x; 
+}
+```
+
+There are several potential choices to we could make here:
+
+1.  We could say that any identifier that can be resolved before macro
+    application keeps its original resolution (so `x` would still resolve to the
+    original, top level `x`).
+2.  We could re-resolve all identifiers after the macros are applied, which can
+    possibly change what they resolve to (in this case `x` would resolve to the
+    generated instance getter `x`).
+3.  We could make it some kind of error for a macro to introduce an identifier
+    that shadows another.
+4.  We could make it error to use an identifier shadowed by one produced by a
```

Comment:
"it error" -> "a compile-time error"

Maybe emphasize "use" (`*use*`) to make it clearer how this differs from the previous point.